ext/afform/core/Civi/Afform/PageTokenCredential.php

```diff
@@ -0,0 +1,189 @@
+        }
+
+        // Not a valid AuthX JWT. Proceed to check any other token sources.
+      }
+    }
+
+  }
+
+  /**
+   * When processing CRM_Core_Invoke, check to see if our token allows us to handle this request.
+   *
+   * @param string $route
+   * @param array $jwt
+   * @return bool
+   * @throws \CRM_Core_Exception
+   * @throws \Civi\API\Exception\UnauthorizedException
+   */
+  public function checkAllowedRoute(string $route, array $jwt): bool {
+    $allowedForm = $jwt['afform'];
+
+    $ajaxRoutes = $this->getAllowedRoutes();
+    foreach ($ajaxRoutes as $regex => $routeInfo) {
+      if (preg_match($regex, $route)) {
+        $parsed = json_decode(\CRM_Utils_Request::retrieve('params', 'String'), 1);
+        if (empty($parsed)) {
+          \Civi::log()->warning("Malformed request. Routes matching $regex must submit params as JSON field.");
+          return FALSE;
+        }
+
+        $extraFields = array_diff(array_keys($parsed), $routeInfo['allowFields']);
+        if (!empty($extraFields)) {
+          \Civi::log()->warning("Malformed request. Routes matching $regex only support these input fields: " . json_encode($routeInfo['allowFields']));
+          return FALSE;
+        }
+
+        $actualForm = $parsed[$routeInfo['nameField']] ?? NULL;
+        if ($actualForm !== $allowedForm) {
+          \Civi::log()->warning("Malformed request. Requested form ($actualForm) does not match allowed name ($allowedForm).");
+          return FALSE;
+        }
+
+        return TRUE;
+      }
+    }
+
+    // Actually, we may not need this? aiming for model where top page-request auth is irrelevant to subrequests...
+    $allowedFormRoute = \Civi\Api4\Afform::get(FALSE)->addWhere('name', '=', $allowedForm)
+      ->addSelect('name', 'server_route')
+      ->execute()
+      ->single();
+    if ($allowedFormRoute['server_route'] === $route) {
+      return TRUE;
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * @return array[]
+   */
+  protected function getAllowedRoutes(): array {
+    return [
+      // ';civicrm/path/to/some/page;' => [
+      //
+      //    // All the fields that are allowed for this API call.
+      //    // N.B. Fields like "chain" are NOT allowed.
+      //    'allowFields' => ['field_1', 'field_2', ...]
+      //
+      //    // The specific field which identifies the target form. Must match our page-token.
+      //    'nameField' => 'field_1',
+      //
+      // ],
+
+      ';^civicrm/ajax/api4/Afform/prefill$;' => [
+        'allowFields' => ['name', 'args'],
+        'nameField' => 'name',
+      ],
+      ';^civicrm/ajax/api4/Afform/submit$;' => [
+        'allowFields' => ['name', 'args', 'values'],
+        'nameField' => 'name',
+      ],
+      ';^civicrm/ajax/api4/Afform/submitFile$;' => [
+        'allowFields' => ['name'], /* FIXME */
+        'nameField' => 'name',
+      ],
+      ';^civicrm/ajax/api4/\w+/autocomplete$;' => [
+        'allowFields' => ['formName'], /* FIXME */
```

Comment:
needs `input`, `page` and `fieldName` at least

Reply:
Brilliant. These will be included in the next push.

(Note: Ben and I also chatted with Coleman, and we'll add few more beyond that.)